firebase-firestore/src/main/java/com/google/firebase/firestore/auth/FirebaseAuthCredentialsProvider.java

```diff
@@ -60,30 +64,47 @@
   private boolean forceRefresh;
 
   /** Creates a new FirebaseAuthCredentialsProvider. */
-  public FirebaseAuthCredentialsProvider(InternalAuthProvider authProvider) {
-    this.authProvider = authProvider;
+  public FirebaseAuthCredentialsProvider(Deferred<InternalAuthProvider> deferredAuthProvider) {
     this.idTokenListener =
         token -> {
           synchronized (this) {
-            currentUser = getUser();
+            currentUser = getUser(internalAuthProviderProviderRef.get());
             tokenCounter++;
 
             if (changeListener != null) {
               changeListener.onValue(currentUser);
             }
           }
         };
-    currentUser = getUser();
+
+    internalAuthProviderProviderRef = new AtomicReference<>(() -> null);
+    currentUser = getUser(internalAuthProviderProviderRef.get());
     tokenCounter = 0;
 
-    authProvider.addIdTokenListener(idTokenListener);
+    deferredAuthProvider.whenAvailable(
+        provider -> {
+          synchronized (this) {
+            internalAuthProviderProviderRef.set(provider);
+            currentUser = getUser(internalAuthProviderProviderRef.get());
+            if (changeListener != null) {
+              changeListener.onValue(currentUser);
```

Comment:
`whenAvailable` will run on the UI thread, is it ok to execute the changeListener on it?